express_server.js

```diff
@@ -6,135 +6,175 @@ const cookieSession = require("cookie-session");
 const bcrypt = require("bcrypt");
 const methodOverride = require("method-override");
 // helper functions:
-const { generateID, getUserByEmail, userURLs, copy } = require("./helpers.js");
+const {
+  areValidEmailPassword,
+  generateID,
+  getUserByEmail,
+  hashPassword,
+  userURLs,
+  emailExists,
+} = require("./helpers.js");
 
 app.set("view engine", "ejs");
-app.use(bodyParser.urlencoded({extended: true}));
-app.use(cookieSession({
-  name: "session",
-  keys: ["user_id"]
-}));
-app.use(methodOverride('_method'));
+app.use(bodyParser.urlencoded({ extended: true }));
+app.use(
+  cookieSession({
+    name: "session",
+    keys: ["user_id"],
+  }),
+);
+app.use(methodOverride("_method"));
 
 // used to keep track of all the URLs and their shortened forms
 const urlDatabase = {
-  "b2xVn2": { longURL: "http://www.lighthouselabs.ca", userID: "dummy" },
-  "icon": { longURL: "https://i.pinimg.com/originals/b9/3f/80/b93f8070c4178d1906198e264cb4c98f.png", userID: "dummy" }
+  b2xVn2: { longURL: "http://www.lighthouselabs.ca", userID: "dummy" },
+  icon: {
+    longURL:
+      "https://i.pinimg.com/originals/b9/3f/80/b93f8070c4178d1906198e264cb4c98f.png",
+    userID: "dummy",
+  },
 };
 
 //global variable, used to store and access users in the app
-let users = {  };
+const users = {};
+
+// Create an endpoint that renders an error info page.
+app.get("/errors/:statusCode", (req, res) => {
+  templateVars = {
+    statusCode: req.params.statusCode,
+    user: req.session.id,
+  };
+  res.render("url_errors", templateVars);
+});
```

Comment:
Here I'm creating an endpoint to render errors. In this way you can organize better the different error messages and have them nicely presented in the UI.